compiler/noirc_frontend/src/hir/comptime/interpreter/builtin.rs

```diff
@@ -2771,3 +2774,54 @@ fn ctstring_eq(arguments: Vec<(Value, Location)>, location: Location) -> IResult
 fn ctstring_hash(arguments: Vec<(Value, Location)>, location: Location) -> IResult<Value> {
     hash_item(arguments, location, get_ctstring)
 }
+
+fn derive_generators(
+    interner: &mut NodeInterner,
+    arguments: Vec<(Value, Location)>,
+    return_type: Type,
+    location: Location,
+) -> IResult<Value> {
+    assert_eq!(arguments.len(), 2);
+    let (domain_separator_string, starting_index) = check_two_arguments(arguments, location)?;
+
+    let domain_separator_location = domain_separator_string.1;
+    let (domain_separator_string, _) = get_array(interner, domain_separator_string)?;
+    let starting_index = get_u32(starting_index)?;
+
+    let domain_separator_string =
+        try_vecmap(domain_separator_string, |byte| get_u8((byte, domain_separator_location)))?;
+
+    let (size, elements) = match return_type.clone() {
+        Type::Array(size, elements) => (size, elements),
+        _ => panic!("ICE: Should only have an array return type"),
+    };
+
+    let Some(num_generators) = size.evaluate_to_u32() else {
+        return Err(InterpreterError::UnknownArrayLength { length: *size, location });
+    };
+
+    let generators = bn254_blackbox_solver::derive_generators(
+        &domain_separator_string,
+        num_generators,
+        starting_index,
+    );
+
+    let is_infinite = FieldElement::zero();
+    let mut results = Vector::new();
+    for gen in generators {
+        let x_big: BigUint = gen.x.into();
+        let x = FieldElement::from_be_bytes_reduce(&x_big.to_bytes_be());
+        let y_big: BigUint = gen.y.into();
+        let y = FieldElement::from_be_bytes_reduce(&y_big.to_bytes_be());
+        let mut embedded_curve_point_fields = HashMap::default();
+        embedded_curve_point_fields.insert(Rc::from("x".to_owned()), Value::Field(x));
+        embedded_curve_point_fields.insert(Rc::from("y".to_owned()), Value::Field(y));
```

Comment:
nitpick: We could create the names of x and y outside the loop then clone them inside to avoid allocating on each iteration

Reply:
done